File: Frontend/trackr-administrador/src/app/shared/http/catalogo/punto-venta.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { PuntoVenta } from '@models/catalogo/punto-venta';
import { Usuario } from '@models/seguridad/usuario';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class PuntoVentaService {
  private dataUrl = 'puntoVenta/';

  constructor(public http: HttpClient) {}

 public consultarTodosParaGrid(): Observable<PuntoVenta[]> {
    return this.http.get<PuntoVenta[]>(this.dataUrl + 'consultarTodosParaGrid');
  }

  public consultarTodosParaSelector(): Observable<PuntoVenta[]> {
    return this.http.get<PuntoVenta[]>(this.dataUrl + 'consultarTodosParaSelector');
  }

  public consultarPorUsuarioEnSesion(): Observable<PuntoVenta> {
    return this.http.get<PuntoVenta>(this.dataUrl + 'consultarPorUsuarioEnSesion');
  }

  public consultar(idPuntoVenta: number): Observable<PuntoVenta> {
    return this.http.get<PuntoVenta>(this.dataUrl + `consultar/${idPuntoVenta}`);
  }

  public consultarUsuariosAsignados(idPuntoVenta: number): Observable<Usuario[]> {
    return this.http.get<Usuario[]>(this.dataUrl + `consultarUsuariosAsignados/${idPuntoVenta}`);
  }

  public agregar(puntoVenta: PuntoVenta): Observable<void> {
    return this.http.post<void>(this.dataUrl + 'agregar', puntoVenta);
  }

  public editar(puntoVenta: PuntoVenta): Observable<void> {
    return this.http.put<void>(this.dataUrl + 'editar', puntoVenta);
  }

  public eliminar(idPuntoVenta: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idPuntoVenta}`);
  }

}

File: Backend/TrackrAPI/TrackrAPI/Services/Catalogo/PuntoVentaService.cs
using TrackrAPI.Dtos.Catalogo;
using TrackrAPI.Dtos.Seguridad;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Catalogo;
using System.Collections.Generic;

namespace TrackrAPI.Services.Catalogo
{
    public class PuntoVentaService
    {

        private IPuntoVentaRepository puntoVentaRepository;
        private PuntoVentaValidatorService puntoVentaValidatorService;

        public PuntoVentaService(IPuntoVentaRepository puntoVentaRepository,
            PuntoVentaValidatorService puntoVentaValidatorService)
        {
            this.puntoVentaValidatorService = puntoVentaValidatorService;
            this.puntoVentaRepository = puntoVentaRepository;
        }

        public PuntoVentaDto ConsultarDto(int idPuntoVenta)
        {
            var puntoVenta = puntoVentaRepository.ConsultarDto(idPuntoVenta);
            puntoVentaValidatorService.ValidarExistencia(puntoVenta);
            return puntoVenta;
        }

        public PuntoVentaDto ConsultarDto(int idPuntoVenta, Usuario usuarioSesion)
        {
            var puntoVenta = puntoVentaRepository.ConsultarDto(idPuntoVenta);
            puntoVentaValidatorService.ValidarExistencia(puntoVenta);
            puntoVenta.NombreVendedor = usuarioSesion.Nombre + " " + usuarioSesion.ApellidoPaterno + " " + (usuarioSesion.ApellidoMaterno != null ? usuarioSesion.ApellidoMaterno : " ");
            puntoVenta.IdVendedor = usuarioSesion.IdUsuario;
            return puntoVenta;
        }

        public IEnumerable<PuntoVentaGridDto> ConsultarTodosParaGrid(int idCompania)
        {
            return puntoVentaRepository.ConsultarTodosParaGrid(idCompania);
        }

        public IEnumerable<PuntoVentaDto> ConsultarTodosParaSelector(int idCompania)
        {
            return puntoVentaRepository.ConsultarTodosParaSelector(idCompania);
        }

        public PuntoVentaDto consultarPorUsuarioEnSesion(Usuario usuarioSesion)
        {
            if(usuarioSesion.IdPuntoVenta != null)
                return puntoVentaRepository.ConsultarDto((int)usuarioSesion.IdPuntoVenta);
            else
                return new PuntoVentaDto{ IdPuntoVenta = 0};
        }

        public IEnumerable<UsuarioDto> ConsultarUsuariosAsignados(int idPuntoVenta)
        {
            return puntoVentaRepository.ConsultarUsuariosAsignados(idPuntoVenta);
        }

        public int Agregar(PuntoVenta puntoVenta)
        {
            puntoVentaValidatorService.ValidarAgregar(puntoVenta);
            puntoVentaRepository.Agregar(puntoVenta);
            return puntoVenta.IdPuntoVenta;
        }

        public void Editar(PuntoVenta tipoPresentacion)
        {
            puntoVentaValidatorService.ValidarEditar(tipoPresentacion);
            puntoVentaRepository.Editar(tipoPresentacion);
        }

        public void Eliminar(int idPuntoVenta)
        {
            var puntoVenta = puntoVentaRepository.Consultar(idPuntoVenta);
            puntoVentaValidatorService.ValidarEliminar(idPuntoVenta);
            puntoVentaRepository.Eliminar(puntoVenta);
        }

    }
}
